### ISSUE TITLE
Using IScriptCustomFunction with Invoke and InvokeAsync

### ISSUE THREAD

**Issue by pebezo** [scriban/scriban#399]
Hi,

We have some old code that has a bunch of custom functions that implement IScriptCustomFunction. After upgrading to the latest Scriban version, this interface now has both Invoke() and InvokeAsync().

Should every custom function implement both? Is there any deterministic way of knowing which one is going to be called?

If I'm looking at the [IncludeFunction](https://github.com/scriban/scriban/blob/master/src/Scriban/Functions/IncludeFunction.cs#L29) it's implementing only the Invoke(). Any way to do that same from external code?

Thanks

**Comment by xoofx** [comment]
> Should every custom function implement both? Is there any deterministic way of knowing which one is going to be called?

If you call RenderAsync from top level Template. It propagates.

> If I'm looking at the IncludeFunction it's implementing only the Invoke(). Any way to do that same from external code?

`IncludeFunction` implements also `InvokeAsync` but it is generated automatically.

> Any way to do that same from external code?

If your code doesn't care about Async, make the async returning the result of the sync method, something like:

```c#
        ValueTask<object> InvokeAsync(TemplateContext context, ScriptNode callerContext, ScriptArray arguments, ScriptBlockStatement blockStatement) {
             return ValueTask<object>.FromResult(Invoke(...));
        }
```

Could be implemented with a default interface method on your side if you can afford the runtime requirements (create an interface inheriting from `IScriptCustomFunction` and implement InvokeAsync similar to the above)

**Comment by pebezo** [comment]
Thanks, that helps! 

In one of the custom functions we're allowing a variable number of parameters, something like:

`{{ my_function 'p0' 'p1' 'p2' 'p3' 'p4' ... 'pN' }}`

The first two parameters (p0 and p1) are always required, the rest are optional. For this scenario, what should `RequiredParameterCount` and `ParameterCount` be set to? 

Scriban already knows the number of parameters (ScriptArray.Count)... why do I have to tell it again?

**Comment by xoofx** [comment]
> Scriban already knows the number of parameters (ScriptArray.Count)... why do I have to tell it again?

This is for the calling site. Not the function called.


This part is not well documented, as I developed it for [kalk](https://kalk.dev) as now with Scriban you can introduce [named parameters and optional values](https://github.com/scriban/scriban/blob/master/doc/language.md#73-parametric-functions)

For example:

```
func sub_opt(x, y, z = 1, w = 2)
   ret x - y - z - w
end
```

- `RequiredParameterCount` here would be = `2`
- `ParameterCount` would be = `4`

I'm curious though why you did use `IScriptCustomFunction` and not delegates? `IScriptCustomFunction` is very low level and I have only used it for some very specific use cases.

**Comment by pebezo** [comment]
> I'm curious though why you did use IScriptCustomFunction and not delegates?

We have over 40 functions built around `IScriptCustomFunction` that started before Scriban reached 1.0. At the time this approach gave us the most amount of flexibility.

One of the reasons I remember was that we could control the exact wording of error messages. We could give hints/suggestions on what to check or how to fix the issue. Also, I could be wrong, but I don't think the line number + column was part of the standard error message at the time. 

We used the SourceSpan to output this info which is super useful when dealing with large templates. Now that I look back, we ended up building similar functionality to what you built in the newer versions.

For example, we needed a way to support a variable number of arguments so we just extracted those from `ScriptArray`. Or we needed a way to make argument N+1 mandatory if the value of argument N was X.

With the current version, we set `RequiredParameterCount` to be the same as `ParameterCount` and dynamically extract the optional ones where needed. This seems to cover all our test cases.

Thanks for your time and building Scriban. All our tests now pass with version 5!
